crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -46,10 +46,23 @@ impl<'db> UnionBuilder<'db> {
     pub(crate) fn add(mut self, ty: Type<'db>) -> Self {
         match ty {
             Type::Union(union) => {
-                self.elements.extend(union.elements(self.db));
+                for element in union.elements(self.db) {
```

Comment:
Nit: It could be worth to call `union.reserve` here based on the assumption that most elements aren't overlapping